crates/sui-source-validation/src/lib.rs

```diff
@@ -0,0 +1,252 @@
+            if pkg_symbol == compiled_package.compiled_package_info.package_name {
+                continue;
+            };
+
+            // fetch the Sui object at the address specified for the package in the local resolution table
+            let on_chain_package = self.pkg_for_address(&addr).await?;
+
+            for (module_symbol, local_bytes) in local_pkg_bytes {
+                // package addresses may show up many times, but we only need to verify them once
+                // zero address is the package we're checking dependencies for
+                if verified_dependencies.contains_key(&addr) || addr == AccountAddress::ZERO {
+                    continue;
+                }
+
+                let mod_str = module_symbol.to_string();
+                let on_chain_bytes = match on_chain_package.module_map.get(&mod_str) {
+                    Some(oc_bytes) => oc_bytes,
+                    None => {
+                        return Err(DependencyVerificationError::OnChainDependencyNotFound(
+                            pkg_symbol,
+                            Some(module_symbol),
+                        ))
+                    }
+                };
+
+                // compare local bytecode to on-chain bytecode to ensure integrity of our dependencies
+                if &local_bytes != on_chain_bytes {
+                    return Err(DependencyVerificationError::ModuleBytecodeMismatch(
+                        pkg_symbol.to_string(),
+                        module_symbol.to_string(),
+                        addr,
+                    ));
+                }
+
+                on_chain_module_count += 1;
+
+                if self.verbose {
+                    println!(
+                        "{}::{} - {} bytes, code matches",
+                        pkg_symbol,
+                        module_symbol,
+                        on_chain_bytes.len()
+                    );
+                }
+            }
+
+            verified_dependencies.insert(
+                addr,
+                Dependency {
+                    symbol: pkg_symbol.to_string(),
+                    module_bytes: on_chain_package.module_map.clone(),
+                },
+            );
+        }
+
+        // total number of modules in packages must match, in addition to each individual module matching
+        let len = compiled_package.deps_compiled_units.len();
+        // only need to check for less than, because if local modules are missing on-chain we've already errored out
+        if len < on_chain_module_count {
+            let missing_modules =
+                Self::get_missing_modules(&compiled_package, &verified_dependencies);
+            return Err(DependencyVerificationError::ModuleCountMismatch(
+                len,
+                on_chain_module_count,
+                missing_modules,
+            ));
+        }
+
+        Ok(DependencyVerificationResult {
+            verified_dependencies,
```

Comment:
I didn't spot a place where we use this return value currently.  Do we expect to in the near future?  If not, I'd prefer to just return `()` here.

Reply:
Related, I think the underlying `verified_dependencies` data structure (currently a map: `AccountAddress -> (PackageName, ModuleName -> ModuleBytes)`) is playing triple duty here, perhaps because it needs to make sense as a return value which may not be necessary, and the code might be simplified by separately tracking:

- A set of visited addresses in a `HashSet` (you can seed this with `AccountAddress::ZERO` to avoid special casing it).
- A set of visited modules (i.e. a `HashSet<(Symbol, Symbol)>`) to later detect missing modules with.

And not cloning the module bytes after verification, because I couldn't spot a reason for `verified_dependencies` to include it.

Reply:
> I didn't spot a place where we use this return value currently. Do we expect to in the near future? If not, I'd prefer to just return `()` here.

no we don't directly use it yet, however it seems preferable to keep some summary of what was checked rather than only offering `()`

> the code might be simplified by separately tracking: 
* A set of visited addresses in a HashSet (you can seed this with AccountAddress::ZERO to avoid special casing it).
* A set of visited modules (i.e. a HashSet<(Symbol, Symbol)>) to later detect missing modules with.

Usually, i find patterns where you're tracking visits less easy to understand, maybe this will shorten the code but i don't prefer it as a mental model